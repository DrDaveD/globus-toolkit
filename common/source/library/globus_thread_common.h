/******************************************************************************
globus_common_thread

Description:

CVS Information:
******************************************************************************/

#ifndef GLOBUS_THREAD_COMMON_H
#define GLOBUS_THREAD_COMMON_H

#include "globus_common_include.h"
#include "globus_module.h"

EXTERN_C_BEGIN


extern globus_module_descriptor_t       globus_i_thread_common_module;

#define GLOBUS_THREAD_COMMON_MODULE     (&globus_i_thread_common_module)

/******************************************************************************
			     Include header files
******************************************************************************/

typedef int                                   globus_thread_result_t;
typedef int                                   globus_thread_callback_index_t;

/**************************************************************************
*  function prototypes
**************************************************************************/
typedef
void *
(*globus_thread_blocking_func_t)(
    globus_thread_callback_index_t      ndx,
    void *                              user_args);

globus_thread_result_t
globus_thread_blocking_callback_push(
    globus_thread_blocking_func_t        func,
    void *                               user_args,
    globus_thread_callback_index_t *     i);

globus_thread_result_t
globus_thread_blocking_callback_pop(
    globus_thread_callback_index_t *     i);

globus_thread_result_t 
globus_thread_blocking_callback_enable(
    globus_thread_callback_index_t *  i);


globus_thread_result_t 
globus_thread_blocking_callback_disable(
    globus_thread_callback_index_t *  i);

globus_thread_result_t 
globus_thread_blocking_will_block();

void
globus_thread_prefork();

void
globus_thread_postfork();

void
globus_thread_blocking_reset();

void thread_print(char * s, ...);

EXTERN_C_END

#endif

