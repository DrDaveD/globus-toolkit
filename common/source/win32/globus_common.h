/*
 * Copyright 1999-2006 University of Chicago
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/******-*- C -*-**************************************************************
globus_common.h

Description:
  Headers common to all of Globus

CVS Information:

  $Source$
  $Date$
  $State$
  $Author$
******************************************************************************/

#if !defined(GLOBUS_INCLUDE_GLOBUS_COMMON_H)
#define GLOBUS_INCLUDE_GLOBUS_COMMON_H 1


/******************************************************************************
		      Include globus_common header files
******************************************************************************/
#include "globus_common_include.h"
#include "globus_args.h"
#include "globus_callback.h"
#include "globus_debug.h"
#include "globus_error.h"
#include "globus_error_errno.h"
#include "globus_error_generic.h"
#include "globus_error_hierarchy.h"
#include "globus_error_string.h"
#include "globus_fifo.h"
#include "globus_handle_table.h"
#include "globus_hashtable.h"
#include "globus_libc.h"
#include "globus_list.h"
#include "globus_memory.h"
#include "globus_module.h"
#include "globus_object.h"
#include "globus_object_cache.h"
#include "globus_object_hierarchy.h"
#include "globus_print.h"
#include "globus_priority_q.h"
#include "globus_strptime.h"
#include "globus_symboltable.h"
#include "globus_thread_common.h" 
#include GLOBUS_THREAD_INCLUDE
#include "globus_thread_pool.h"
#include "globus_tilde_expand.h"
#include "globus_time.h"
#include "globus_url.h"

# if !defined(alloca)
/* AIX requires this to be the first thing in the file.  */
#ifdef __GNUC__
# define alloca __builtin_alloca
#else
# if HAVE_ALLOCA_H
#  include <alloca.h>
# else
#  ifdef _AIX
#pragma alloca
#  else
#   ifndef alloca /* predefined by HP cc +Olibcalls */
#     ifndef _CRAYT3E
char *alloca ();
#     endif
#   endif
#  endif
# endif
#endif
#endif

#if !defined(MAXPATHLEN) 
#   include <sys/param.h>
#   define MAXPATHLEN PATH_MAX
#endif
EXTERN_C_BEGIN

/* most network-related functions (getpeername, getsockname,...) have
   an int* as argument, except AIX which uses size_t*. This will
   masquerade the difference. */
#define globus_netlen_t int

/*
 * globus_barrier_t
 *
 * A generic barrier structure */
typedef struct globus_barrier_s
{
    globus_mutex_t      mutex;
    globus_cond_t       cond;
    int                 count;
} globus_barrier_t;

/******************************************************************************
			       Define constants
******************************************************************************/
 
/******************************************************************************
			  Module activation structure
******************************************************************************/
extern globus_module_descriptor_t	globus_i_common_module;

#define GLOBUS_COMMON_MODULE (&globus_i_common_module)


/******************************************************************************
		   Install path discovery functions
******************************************************************************/

globus_result_t
globus_location (  char **   bufp );

/* returns value of GLOBUS_LOCATION in the deploy dir config file */
globus_result_t
globus_common_get_attribute_from_config_file( 
	char *                                         deploy_path,
    char *                                         file_location,
	char *                                         attribute,
	char **                                        value);

EXTERN_C_END

#endif /* GLOBUS_INCLUDE_GLOBUS_COMMON_H */
