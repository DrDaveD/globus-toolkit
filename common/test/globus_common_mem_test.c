/*
 * Portions of this file Copyright 1999-2005 University of Chicago
 * Portions of this file Copyright 1999-2005 The University of Southern California.
 *
 * This file or a portion of this file is licensed under the
 * terms of the Globus Toolkit Public License, found at
 * http://www.globus.org/toolkit/download/license.html.
 * If you redistribute this file, with or without
 * modifications, you must include this notice in the file.
 */

#include "globus_common.h"
#include "globus_thread_common.h"

static globus_memory_t                   mem;

#define MEM_INIT_SIZE      15
#ifndef TARGET_ARCH_WIN32
#define POPS               100000
#else
#define POPS                 1000
#endif

typedef struct mem_test_s
{
    char a;
    char b;
    char c;
    char x[4];
    
} mem_test_t;

int mem_init(mem_test_t * m, int cnt);

void dump(globus_byte_t * buf, int size);

int main(int argc, char * argv[])
{
   int                         rc = GLOBUS_SUCCESS;
   mem_test_t *                mem_ptr[POPS];
   int                         cnt = 0;

   globus_module_activate(GLOBUS_COMMON_MODULE);

   globus_memory_init(&mem,
		      sizeof(mem_test_t),
		      MEM_INIT_SIZE);

   for(cnt = 0; cnt < POPS; cnt++)
   {
       mem_ptr[cnt] = ( mem_test_t *) globus_memory_pop_node(&mem); 
       mem_init(mem_ptr[cnt], cnt);
   }

   /* nodes are aligned on mod 8 boundaries, add 1 to 7 to make 8 */
   dump((globus_byte_t *) mem_ptr[0], MEM_INIT_SIZE * (sizeof(mem_test_t) + 1));
   
   globus_memory_push_node(&mem, (globus_byte_t *)mem_ptr[0]);
   
   for(cnt = 1; cnt < POPS; cnt++)
   {
       globus_memory_push_node(&mem, (globus_byte_t *) mem_ptr[cnt]);
   }

   globus_memory_destroy(&mem);

   globus_module_deactivate(GLOBUS_COMMON_MODULE);

   return rc;
}

int
mem_init(mem_test_t * m, int cnt)
{
    memset(m, 0, sizeof(mem_test_t) + 1);
    m->a = 'a';
    m->b = 'b';
    m->c = 'c';
 
    return GLOBUS_TRUE;
}

void dump(globus_byte_t * buf, int size)
{
    int  ctr;
    int  col = 0;
    
    printf("printing %d bytes\n", size);
    printf("\n    ....+....1....+....2....+....3....+....4\n    ");
    for(ctr = 0; ctr < size; ctr++)
    {
	printf("%c", isprint(buf[ctr]) ? buf[ctr] : '.');
	col++;
        if(col >= 40)
	{
            printf("\n    ");
	    col = 0;
	}
    }
    printf("\n");
}
