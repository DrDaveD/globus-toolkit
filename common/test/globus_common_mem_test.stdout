printing 120 bytes

    ....+....1....+....2....+....3....+....4
    abc.....abc.....abc.....abc.....abc.....
    abc.....abc.....abc.....abc.....abc.....
    abc.....abc.....abc.....abc.....abc.....
    
